File: internal/service/data_svc/mall_weather_capacity_plan.go
package data_svc

import (
	"errors"
	"math"
)

const (
	defaultMallWeatherCapacityMinutelyRows = 120
	defaultMallWeatherCapacityHourlyBatch  = 200
	defaultMallWeatherCapacityAlertBatch   = 200
	maxMallWeatherCapacityMallCount        = 100000
	maxMallWeatherCapacityProviderQPS      = 10000
	maxMallWeatherCapacityAlertsPerMall    = 256
)

var ErrMallWeatherInvalidCapacityPlan = errors.New("mall weather capacity plan: invalid input")

type MallWeatherCapacityPlanInput struct {
	MallCount       int
	ProviderQPS     float64
	HourlySteps     int
	DailySteps      int
	LifeIndexDays   int
	AlertsPerMall   int
	FeishuBatchRows int
}

type MallWeatherCapacityDatasetPlan struct {
	Kind            string `json:"kind"`
	Rows            int64  `json:"rows"`
	DatabaseBatches int64  `json:"databaseBatches"`
	FeishuBatches   int64  `json:"feishuBatches"`
}

type MallWeatherCapacityPlan struct {
	MallCount                    int                              `json:"mallCount"`
	ProviderQPS                  float64                          `json:"providerQps"`
	ProviderRequests             int64                            `json:"providerRequests"`
	ProviderDrainSeconds         float64                          `json:"providerDrainSeconds"`
	MinimumQPSForHourlyFullSweep float64                          `json:"minimumQpsForHourlyFullSweep"`
	TotalDatabaseRows            int64                            `json:"totalDatabaseRows"`
	TotalDatabaseBatches         int64                            `json:"totalDatabaseBatches"`
	FeishuBatchRows              int                              `json:"feishuBatchRows"`
	TotalFeishuBatches           int64                            `json:"totalFeishuBatches"`
	Datasets                     []MallWeatherCapacityDatasetPlan `json:"datasets"`
}

func BuildMallWeatherCapacityPlan(input MallWeatherCapacityPlanInput) (MallWeatherCapacityPlan, error) {
	normalized, err := normalizeMallWeatherCapacityPlanInput(input)
	if err != nil {
		return MallWeatherCapacityPlan{}, err
	}
	malls := int64(normalized.MallCount)
	providerRequests := malls * 2
	datasets := []MallWeatherCapacityDatasetPlan{
		mallWeatherCapacityDataset("realtime", malls, malls, normalized.FeishuBatchRows),
		mallWeatherCapacityDataset("minutely", malls*defaultMallWeatherCapacityMinutelyRows, malls, normalized.FeishuBatchRows),
		mallWeatherCapacityDataset("hourly", malls*int64(normalized.HourlySteps), malls*ceilInt64(int64(normalized.HourlySteps), defaultMallWeatherCapacityHourlyBatch), normalized.FeishuBatchRows),
		mallWeatherCapacityDataset("daily", malls*int64(normalized.DailySteps), malls, normalized.FeishuBatchRows),
		mallWeatherCapacityDataset("alerts", malls*int64(normalized.AlertsPerMall), malls*ceilInt64(int64(normalized.AlertsPerMall), defaultMallWeatherCapacityAlertBatch), normalized.FeishuBatchRows),
		mallWeatherCapacityDataset("life_indices", malls*int64(normalized.LifeIndexDays), malls, normalized.FeishuBatchRows),
	}
	plan := MallWeatherCapacityPlan{
		MallCount:                    normalized.MallCount,
		ProviderQPS:                  normalized.ProviderQPS,
		ProviderRequests:             providerRequests,
		ProviderDrainSeconds:         float64(providerRequests) / normalized.ProviderQPS,
		MinimumQPSForHourlyFullSweep: float64(providerRequests) / 3600,
		FeishuBatchRows:              normalized.FeishuBatchRows,
		Datasets:                     datasets,
	}
	for _, dataset := range datasets {
		plan.TotalDatabaseRows += dataset.Rows
		plan.TotalDatabaseBatches += dataset.DatabaseBatches
		plan.TotalFeishuBatches += dataset.FeishuBatches
	}
	return plan, nil
}

func normalizeMallWeatherCapacityPlanInput(input MallWeatherCapacityPlanInput) (MallWeatherCapacityPlanInput, error) {
	if input.FeishuBatchRows == 0 {
		input.FeishuBatchRows = defaultMallWeatherFeishuBatchRows
	}
	if input.MallCount < 1 || input.MallCount > maxMallWeatherCapacityMallCount ||
		input.ProviderQPS <= 0 || math.IsNaN(input.ProviderQPS) || math.IsInf(input.ProviderQPS, 0) ||
		input.ProviderQPS > maxMallWeatherCapacityProviderQPS ||
		input.HourlySteps < 1 || input.HourlySteps > 360 ||
		input.DailySteps < 1 || input.DailySteps > 15 ||
		input.LifeIndexDays < 1 || input.LifeIndexDays > 15 ||
		input.AlertsPerMall < 0 || input.AlertsPerMall > maxMallWeatherCapacityAlertsPerMall ||
		input.FeishuBatchRows < 1 || input.FeishuBatchRows > maxMallWeatherFeishuBatchRows {
		return MallWeatherCapacityPlanInput{}, ErrMallWeatherInvalidCapacityPlan
	}
	return input, nil
}

func mallWeatherCapacityDataset(kind string, rows, databaseBatches int64, feishuBatchRows int) MallWeatherCapacityDatasetPlan {
	return MallWeatherCapacityDatasetPlan{
		Kind:            kind,
		Rows:            rows,
		DatabaseBatches: databaseBatches,
		FeishuBatches:   ceilInt64(rows, int64(feishuBatchRows)),
	}
}

func ceilInt64(value, divisor int64) int64 {
	if value <= 0 {
		return 0
	}
	return (value + divisor - 1) / divisor
}

File: web/src/mallWeatherCapacityPlan.ts
export type MallWeatherCapacityPlanInput = {
  mallCount: string
  providerQps: string
  hourlySteps: string
  dailySteps: string
  lifeIndexDays: string
  alertsPerMall: string
  feishuBatchRows: string
}

export type MallWeatherCapacityDataset = {
  kind: string
  rows: number
  databaseBatches: number
  feishuBatches: number
}

export type MallWeatherCapacityPlan = {
  mallCount: number
  providerQps: number
  weatherV26ProviderRequestsPerDay: number
  lifeIndexV3ProviderRequestsPerDay: number
  providerRequests: number
  providerDrainSeconds: number
  minimumQpsForOneHourDrain: number
  totalDatabaseRows: number
  totalDatabaseBatches: number
  feishuBatchRows: number
  totalFeishuBatches: number
  datasets: MallWeatherCapacityDataset[]
}

export function mallWeatherCapacityPlanPath(input: MallWeatherCapacityPlanInput) {
  const mallCount = positiveInteger(input.mallCount, 1, 100_000)
  const providerQps = positiveNumber(input.providerQps, 10_000)
  const hourlySteps = positiveInteger(input.hourlySteps, 1, 360)
  const dailySteps = positiveInteger(input.dailySteps, 1, 15)
  const lifeIndexDays = positiveInteger(input.lifeIndexDays, 1, 15)
  const alertsPerMall = nonNegativeInteger(input.alertsPerMall, 256)
  const feishuBatchRows = positiveInteger(input.feishuBatchRows, 1, 500)
  const query = new URLSearchParams({
    mallCount: String(mallCount),
    providerQps: String(providerQps),
    hourlySteps: String(hourlySteps),
    dailySteps: String(dailySteps),
    lifeIndexDays: String(lifeIndexDays),
    alertsPerMall: String(alertsPerMall),
    feishuBatchRows: String(feishuBatchRows),
  })
  return `/v1/mall-weather/capacity-plan?${query.toString()}`
}

export function parseMallWeatherCapacityPlan(payload: unknown): MallWeatherCapacityPlan | null {
  const data = envelopeData(payload)
  if (!data || !positiveSafeInteger(data.mallCount) || !positiveFiniteNumber(data.providerQps) ||
    !nonNegativeSafeInteger(data.weatherV26ProviderRequestsPerDay) || !nonNegativeSafeInteger(data.lifeIndexV3ProviderRequestsPerDay) ||
    !nonNegativeSafeInteger(data.providerRequests) || !nonNegativeFiniteNumber(data.providerDrainSeconds) ||
    !nonNegativeFiniteNumber(data.minimumQpsForOneHourDrain) || !nonNegativeSafeInteger(data.totalDatabaseRows) ||
    !nonNegativeSafeInteger(data.totalDatabaseBatches) || !positiveSafeInteger(data.feishuBatchRows) ||
    !nonNegativeSafeInteger(data.totalFeishuBatches) || !Array.isArray(data.datasets)) return null

  const datasets: MallWeatherCapacityDataset[] = []
  for (const item of data.datasets) {
    if (!isRecord(item) || typeof item.kind !== 'string' || !item.kind.trim() || !nonNegativeSafeInteger(item.rows) ||
      !nonNegativeSafeInteger(item.databaseBatches) || !nonNegativeSafeInteger(item.feishuBatches)) return null
    datasets.push({ kind: item.kind, rows: item.rows, databaseBatches: item.databaseBatches, feishuBatches: item.feishuBatches })
  }
  return {
    mallCount: data.mallCount,
    providerQps: data.providerQps,
    weatherV26ProviderRequestsPerDay: data.weatherV26ProviderRequestsPerDay,
    lifeIndexV3ProviderRequestsPerDay: data.lifeIndexV3ProviderRequestsPerDay,
    providerRequests: data.providerRequests,
    providerDrainSeconds: data.providerDrainSeconds,
    minimumQpsForOneHourDrain: data.minimumQpsForOneHourDrain,
    totalDatabaseRows: data.totalDatabaseRows,
    totalDatabaseBatches: data.totalDatabaseBatches,
    feishuBatchRows: data.feishuBatchRows,
    totalFeishuBatches: data.totalFeishuBatches,
    datasets,
  }
}

function positiveInteger(value: string, minimum: number, maximum: number) {
  const parsed = Number(value)
  if (!Number.isSafeInteger(parsed) || parsed < minimum || parsed > maximum) throw new Error('invalid capacity plan input')
  return parsed
}

function nonNegativeInteger(value: string, maximum: number) {
  const parsed = Number(value)
  if (!Number.isSafeInteger(parsed) || parsed < 0 || parsed > maximum) throw new Error('invalid capacity plan input')
  return parsed
}

function positiveNumber(value: string, maximum: number) {
  const parsed = Number(value)
  if (!Number.isFinite(parsed) || parsed <= 0 || parsed > maximum) throw new Error('invalid capacity plan input')
  return parsed
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value)
}

function envelopeData(payload: unknown): Record<string, unknown> | null {
  if (!isRecord(payload) || payload.code !== 0 || !isRecord(payload.data)) return null
  return payload.data
}

function positiveSafeInteger(value: unknown): value is number {
  return typeof value === 'number' && Number.isSafeInteger(value) && value > 0
}

function nonNegativeSafeInteger(value: unknown): value is number {
  return typeof value === 'number' && Number.isSafeInteger(value) && value >= 0
}

function positiveFiniteNumber(value: unknown): value is number {
  return typeof value === 'number' && Number.isFinite(value) && value > 0
}

function nonNegativeFiniteNumber(value: unknown): value is number {
  return typeof value === 'number' && Number.isFinite(value) && value >= 0
}
